pandas/core/groupby/groupby.py

```diff
@@ -4317,5 +4317,8 @@ def _insert_quantile_level(idx: Index, qs: npt.NDArray[np.float64]) -> MultiInde
         codes = [np.repeat(x, nqs) for x in idx.codes] + [np.tile(lev_codes, len(idx))]
         mi = MultiIndex(levels=levels, codes=codes, names=idx.names + [None])
     else:
-        mi = MultiIndex.from_product([idx, qs])
+        lev_codes, lev = Index(qs).factorize()
+        levels = [idx, lev]
+        codes = [np.repeat(range(len(idx)), nqs), np.tile(lev_codes, len(idx))]
+        mi = MultiIndex(levels=levels, codes=codes, names=[idx.name, None])
```

Comment:
I'm seeing a bit of a performance regression here from the use of 64 bit ints in codes. The current path will use int8, int16, int32 when appropriate, resulting in 50% speedup in this op (~200ms on my machine) with the following example:

```
size = 10000
idx = pd.Index(range(size), name='cat1')
qs = np.array([e / size for e in range(size)])
```

Converting the inputs of `np.repeat` and `np.tile` to `int16` in this example performs slightly better than main. I think we should find a way to factorize this using the appropriate number of bits for the dtype, but I'll have to take a deeper look to best figure out how.

Reply:
@Charlie-XIAO this avoids the perf regression:

```
from pandas.core.dtypes.cast import coerce_indexer_dtype

nidx = len(idx)
nqs = len(qs)
lev_codes, lev = Index(qs).factorize()
lev_codes = coerce_indexer_dtype(lev_codes, lev)
idx_codes = coerce_indexer_dtype(np.arange(nidx), idx)
levels = [idx, lev]
codes = [np.repeat(idx_codes, nqs), np.tile(lev_codes, nidx)]
mi = MultiIndex(levels=levels, codes=codes, names=[idx.name, None])
```

The import, if necessary, should happen at the top of the file.

Reply:
@rhshadrach Thank you so much! I will commit these changes ASAP.